### DISCUSSION TITLE
weight of a sample

### DISCUSSION THREAD

**Discussion by max-dev-14** [ultralytics/ultralytics#23441] [General]
I have heard that there is a technique to effectively train a model using an unbalanced dataset. The methodology is as follows:
Let's say there are two abstract categories: cats and dogs. But the number of cats is much less than the number of dogs (the case of an unbalanced dataset). You can set up the learning pipeline so that the sample weight is determined, which affects the gradient. For example, there are 100 cats and 10,000 dogs in the dataset. You indicate that the model should be more attentive to cats and change its weights faster than for dogs. It turns out that the model learns at different rates for different classes, since the model is more sensitive to errors in the cat category.
is this technique implemented in Ultralytics?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @max-dev-14, thank you for your interest in Ultralytics 🚀! This is an automated response, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (like class imbalance / per-class sample weighting), please provide as much information as possible so we can point you to the right built-in options or recommended workflow, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).  
Specifically helpful details:
-Which task/mode are you training (detect/segment/classify) and how are you launching training (Python or CLI from the [Docs](https://docs.ultralytics.com/))?
-Your class counts (e.g., cats vs dogs) and a snippet of your dataset YAML
-Your training command/config and any logs that show the issue

Join the Ultralytics community where it suits you best 🙂. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Yes—Ultralytics already addresses class imbalance in detection primarily via loss re-weighting (e.g. `FocalLoss` / `VarifocalLoss`), which increases the gradient contribution from harder/rarer positives rather than treating every easy negative equally (see the [loss implementations reference](https://docs.ultralytics.com/reference/utils/loss/)). If you specifically need explicit per-class/per-sample weighting or oversampling (your “cats get higher weight than dogs” example), that isn’t a single built-in training flag today, but you can implement it cleanly by customizing the trainer/dataloader (see [Advanced Customization](https://docs.ultralytics.com/usage/engine/)); let me know whether you’re training `detect`, `segment`, or `classify` and your class counts so we can point you to the best approach.

**Reply by max-dev-14** [reply]
My task is detection with 6 classes.

**Comment by Y-T-G** [comment]
You can try weighted dataloader:

https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Reply by max-dev-14** [reply]
It looks interesting, thanks for the idea!

**Reply by glenn-jocher** [reply]
Glad it helps.

**Comment by max-dev-14** [comment]
Firstly, I'd like to thank for the idea with weighted dataloader. Nonetheless, I didn't quite put it exactly about my task (I'm sorry).
Here is a diagram that allows to understand better what I need.

<img width="950" height="610" alt="diagram" src="https://github.com/user-attachments/assets/0b6df99a-bc4f-4e86-8d4c-64658c61f94b" />

Let's say, I have two videos with 2000 and 200 frames and each sample has a weight based on frames count. Train, val, test subsets include subfolders related to a corresponding video. So, the dataloader must know which video (i.e. which subfolder in the dataset) each sample belongs to. All the samples in the batch are fed to the model and the model predicts. Then the predictions are compared with the annotations and loss functions are computed. I need each loss function per sample must be multiplied by respective sample weight before averaging and feeding to back propagation process. Is it implemented in Ultralytics yet?

**Reply by glenn-jocher** [reply]
This exact “multiply each sample’s loss by a per-video weight before reduction” isn’t exposed as a built-in Ultralytics config option yet; to do it you’ll need a small customization where your dataset adds a `weight` field (based on the sample’s parent video/subfolder), and your model/trainer applies that weight inside the criterion before the final mean/sum (the supported hook for this is a custom model via `init_criterion()` as shown in the [Advanced Customization guide](https://docs.ultralytics.com/usage/engine/) and you can also attach/adjust fields in `BaseTrainer.preprocess_batch()` from the [BaseTrainer reference](https://docs.ultralytics.com/reference/engine/trainer/)). If you share which task you’re training (detect/segment/classify) and what `batch` keys you have right before loss, I can point you to the most appropriate place to apply the weighted reduction (the loss building blocks live in the [loss reference](https://docs.ultralytics.com/reference/utils/loss/)).

**Reply by max-dev-14** [reply]
I'm working on a detection task. But I don't quite understand what batch keys you mean.

**Reply by glenn-jocher** [reply]
By “batch keys” I mean the fields in the Python `dict` produced by the dataloader and passed into the model loss; for detection it typically includes `img`, `cls`, `bboxes`, and `batch_idx` (where `batch_idx` tells you which image each label belongs to), as shown in the [model loss reference](https://docs.ultralytics.com/reference/nn/tasks/). To see exactly what you have in your run (and whether you already get an `im_file`/path you can map back to a video subfolder), print `batch.keys()` once like this:

```python
from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer

class DebugTrainer(DetectionTrainer):
    def preprocess_batch(self, batch):
        batch = super().preprocess_batch(batch)
        if not hasattr(self, "_printed_batch"):
            self._printed_batch = True
            print("batch keys:", list(batch.keys()))
            for k, v in batch.items():
                print(k, getattr(v, "shape", type(v)))
        return batch

model = YOLO("yolo26n.pt")
model.train(data="your_dataset.yaml", epochs=1, trainer=DebugTrainer)
```

Paste the printed keys/output here (text, not a screenshot), and confirm whether your per-video weight should be per-image or per-label; that determines whether we index weights directly by image in the batch or expand via `batch_idx` (see the [Detection task documentation](https://docs.ultralytics.com/tasks/detect/) for how detection batches are structured).

**Reply by max-dev-14** [reply]
batch keys: ['batch_idx', 'bboxes', 'cls', 'im_file', 'img', 'ori_shape', 'resized_shape']
batch_idx torch.Size([29])
bboxes torch.Size([29, 4])
cls torch.Size([29, 1])
im_file <class 'list'>
img torch.Size([4, 3, 640, 640])
ori_shape <class 'list'>
resized_shape <class 'list'>

Per-video weight should be per-image

**Reply by glenn-jocher** [reply]
Thanks—those keys are exactly what you need: `im_file` is a per-image list (length `bs`) that tells you which video/subfolder each image came from (it’s populated by the dataset) so you can derive a per-image weight directly from `Path(im_file[i]).parent`. See the [YOLODataset reference](https://docs.ultralytics.com/reference/data/dataset/) for where `im_file` comes from. ([docs.ultralytics.com](https://docs.ultralytics.com/reference/data/dataset/))

The one catch is that detection loss isn’t naturally “per-image” internally (it’s accumulated over labels/assignments), so to get *true* per-image weighting you’d typically map image weights onto labels via `batch_idx` and apply the weights inside a custom criterion before the final reduction; if you’re OK with an equivalent effect in expectation, a simpler approach is to implement per-video weighted sampling (oversample frames from the short video) in a custom dataloader without touching loss.

If you paste 2–3 example `im_file` strings (as text) and your exact rule for converting “video frames count → weight”, I can point you to the cleanest place to inject that (sampler vs. custom criterion) for your setup.
